mojo/core/broker_castanets.cc

```diff
@@ -353,13 +392,62 @@ void BrokerCastanets::SyncSharedBufferImpl(const base::UnguessableToken& guid,
 
 void BrokerCastanets::SyncSharedBufferImpl2d(const base::UnguessableToken& guid,
                                              uint8_t* memory,
-                                             size_t offset,
-                                             size_t sync_size,
                                              size_t mapped_size,
                                              size_t width,
+                                             size_t height,
+                                             size_t bytes_per_pixel,
+                                             size_t offset,
                                              size_t stride,
+                                             BrokerCompressionMode compression_mode,
                                              bool write_lock) {
+  size_t sync_size = width * height * bytes_per_pixel;
   CHECK_GE(mapped_size, offset + sync_size);
+  size_t put_bytes = 0;
+  size_t put_offset = offset;
+  uint32_t buffer_stride = width * bytes_per_pixel;
+
+  std::vector<uint8_t> bytes(sync_size);
+  if (offset == 0 && stride == 0) {
+      // Full tile
+      memcpy(static_cast<uint8_t*>(&bytes[0]) + offset, memory,
```

Comment:
It seems we can compress the memory directly without copy.

Reply:
Updated the patchset without copy.